### PR TITLE
Session Management API enhancements

### PR DESCRIPTION
**Pull request by jcgarciaa**
### Proposed changes in this pull request
Includes new features to the session management API as per wso2/product-is#12193:
- Search active sessions on the system: includes filtering (by user agent, client IP address and login time among other fields) and pagination capabilities.
- Retrieve a specific session for a certain user.

### Related PRs
- wso2/identity-api-user#132
- wso2/identity-rest-dispatcher#390
- wso2/product-is#12249
- wso2/docs-is#2714

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionMgtDBQueries.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            "LISTAGG(APP_ID || '':'' || APP_NAME || '':'' || SUBJECT || '':'' || UUID, ''|'') " +
+            "WITHIN GROUP (ORDER BY APP_ID) AS APPLICATIONS FROM IDN_AUTH_SESSION_APP_INFO ai " +
+            "JOIN APPS a ON a.ID = ai.APP_ID WHERE APP_ID IN (SELECT ID FROM APPS){2} GROUP BY SESSION_ID";
+
+    private static final String LOAD_SESSION_APP_DATA_MSSQL = "SELECT SESSION_ID, " +
+            "STRING_AGG(CONCAT_WS('':'', APP_ID, APP_NAME, SUBJECT, UUID), ''|'') AS APPLICATIONS " +
+            "FROM IDN_AUTH_SESSION_APP_INFO ai JOIN APPS a ON a.ID = ai.APP_ID " +
+            "WHERE APP_ID IN (SELECT ID FROM APPS){2} GROUP BY SESSION_ID";
+
+    public static final String LOAD_SESSIONS_MYSQL = "WITH SESSIONS AS (" + LOAD_SESSION_STORE_DATA + "), " +
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionMgtDBQueries.java:67]: As raised by @vihanga-liyanage 'WITH' keyword will not work in pre mysql 8.0 . Please revise the query to work with mysql versions from 5.7

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionMgtDBQueries.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            "SESSION_APPS AS (" + LOAD_SESSION_APP_DATA_ORACLE + ") SELECT ss.SESSION_ID, TIME_CREATED, USER_ID, " +
+            "IP_ADDRESS, LOGIN_TIME, LAST_ACCESS_TIME, USER_AGENT, APPLICATIONS FROM IDN_AUTH_SESSION_STORE ss " +
+            "JOIN IDN_AUTH_USER_SESSION_MAPPING sm ON sm.SESSION_ID = ss.SESSION_ID " +
+            "JOIN SESSION_METADATA md ON md.SESSION_ID = ss.SESSION_ID " +
+            "JOIN SESSION_APPS sa ON sa.SESSION_ID = ss.SESSION_ID " +
+            "WHERE (ss.SESSION_ID, TIME_CREATED) IN (SELECT SESSION_ID, TIME_CREATED FROM SESSIONS) AND " +
+            "OPERATION = ''STORE'' AND SESSION_TYPE = ''AppAuthFrameworkSessionContextCache''{3} " +
+            "ORDER BY TIME_CREATED {4} FETCH FIRST {5} ROWS ONLY";
+
+    public static final String LOAD_SESSIONS_MSSQL = "WITH SESSIONS AS (" + LOAD_SESSION_STORE_DATA + "), " +
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionMgtDBQueries.java:111]: Please check if the queries are working properly in the relevant databases and in all supported versions (if not already done)

Comment (jcgarciaa) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionMgtDBQueries.java:109]: Tested on H2, MySQL 5.7/8.0, Oracle 12c/19c, Postgres 10.5/12.2/13, DB2 11.5, SQLServer 2017/2019.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionMgtDBQueries.java (hunk 1 above)
Comment (jcgarciaa) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionMgtDBQueries.java]: MySQL query updated to work on 5.7

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/UserSessionManagementService.java [hunk 3]
```diff
@@ -116,4 +131,20 @@ default boolean terminateSessionBySessionId(User user, int idpId, String session
...
+     * @param filter       criteria to search for sessions
+     * @param limit        maximum number of sessions to be returned in the result set
+     * @param sortOrder    sort direction for results (ASC, DESC)
+     * @return list of session search result objects. In the default method, null is returned.
+     * @throws SessionManagementException if the session retrieval fails
+     */
+    default List<SessionSearchResult> getSessions(String tenantDomain, List<ExpressionNode> filter, Integer limit,
+                                                  String sortOrder) throws SessionManagementException {
+
+        return null;
```
[the remark this answers is not in the document]
Comment (jcgarciaa) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/UserSessionManagementService.java:148]: Updated as per suggestion

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/SessionSearchResult.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ * Additional properties include:
+ * <ul>
+ * <li>User ID
+ * <li>Creation time
+ * </ul>
+ */
+public class SessionSearchResult extends UserSession {
+
+    private String userId;
+    private Long creationTime;
```
[the remark this answers is not in the document]
Comment (jcgarciaa) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/SessionSearchResult.java]: Updated as per suggestion

**Review by pulasthi7 — APPROVED** [review]

**Pull request merged**